fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -427,6 +260,167 @@ def _threshold_optimization_equalized_odds(sensitive_features, labels, scores, g
+
+        if not is_classifier(self.estimator):
+            raise ValueError("{} needs a classifier to work on, and {} is not "
+                             "one.".format(self.__class__.__name__,
+                                           self.estimator.__class__.__name__))
+
+        X, y, sensitive_features = self._validate_input_data(
+            X, sensitive_features, y)
+
+        self.estimator_ = self.estimator.fit(X, y, **kwargs)
+
+        scores = _get_soft_predictions(self.estimator_, X)
+        self._post_processed_predictor_by_sensitive_feature = threshold_optimizer(
+            sensitive_features, y, scores, self.grid_size, self.flip, self.plot)
+
+    def predict(self, X, *, sensitive_features):
+        """Predict label for each sample in X while taking into account sensitive features.
+
+        :param X: feature matrix
+        :type X: numpy.ndarray or pandas.DataFrame
+
+        :param sensitive_features: sensitive features to identify groups by, currently allows
+            only a single column
+        :type sensitive_features: currently 1D array as numpy.ndarray, list, pandas.DataFrame,
+            or pandas.Series
+
+        :return: predictions in numpy.ndarray
+        """
+        check_is_fitted(self)
+        random_state = check_random_state(self.random_state)
+
+        X, _, sensitive_features = self._validate_input_data(
+            X, sensitive_features)
+
+        unconstrained_predictions = self._unconstrained_predictor.predict(X)
+
+        positive_probs = _vectorized_prediction(
+            self._post_processed_predictor_by_sensitive_feature,
+            sensitive_features,
+            unconstrained_predictions)
+        return (positive_probs >= random_state.rand(len(positive_probs))) * 1
+
+    def _pmf_predict(self, X, *, sensitive_features):
+        """Probabilistic mass function.
+
+        :param X: Feature matrix
+        :type X: numpy.ndarray or pandas.DataFrame
+
+        :param sensitive_features: Sensitive features to identify groups by,
+            currently allows only a single column
+
+        :type sensitive_features: Currently 1D array as numpy.ndarray, list,
+            pandas.DataFrame, or pandas.Series
+
+        :return: array of tuples with probabilities for predicting 0 or 1,
+            respectively. The sum of the two numbers in each tuple needs to add
+            up to 1.
+        :rtype: numpy.ndarray
+        """
+        check_is_fitted(self)
+        X, _, sensitive_features = self._validate_input_data(
+            X, sensitive_features)
+        positive_probs = _vectorized_prediction(
+            self._post_processed_predictor_by_sensitive_feature, sensitive_features,
+            self._unconstrained_predictor.predict(X))
+        return np.array([[1.0 - p, p] for p in positive_probs])
+
+    def _validate_input_data(self, X, sensitive_features, y=None):
+        if y:
+            X, y = check_X_y(X, y)
```

Comment:
I've looked at this a little bit, but it's not entirely clear to me what types are allowed. We currently support list, dataframe, series, and ndarray for y and sensitive_features, X.

Reply:
yes, all of those are cast into an ndarray, through `np.asarray`.